File: backend/strategy_engine.py
from typing import List, Dict, Optional, Tuple
import pandas as pd
import numpy as np


class StrategyEngine:
    def __init__(self, blocks: List[Dict]):
        self.blocks = blocks
        self.positions: List[Dict] = []
        self.pnl: float = 0
        self.trade_count: int = 0

    def should_enter(self, market_data: pd.DataFrame) -> bool:
        """Check all entry conditions."""
        entry_blocks = [b for b in self.blocks if b.get("category") == "entry"]
        if not entry_blocks:
            return False

        for block in entry_blocks:
            block_id = block.get("id", "")
            params = block.get("params", {})

            if block_id == "rsi-gate":
                rsi = self._calculate_rsi(market_data)
                if rsi > params.get("threshold", 30):
                    return False

            elif block_id == "ma-cross":
                fast = params.get("fast", 9)
                slow = params.get("slow", 21)
                fast_ma = market_data["close"].rolling(fast).mean()
                slow_ma = market_data["close"].rolling(slow).mean()
                if fast_ma.iloc[-1] <= slow_ma.iloc[-1]:
                    return False

            elif block_id == "macd-signal":
                fast = params.get("fast", 12)
                slow = params.get("slow", 26)
                fast_ema = market_data["close"].ewm(span=fast).mean()
                slow_ema = market_data["close"].ewm(span=slow).mean()
                macd = fast_ema.iloc[-1] - slow_ema.iloc[-1]
                if macd <= 0:
                    return False

        return True

    def calculate_position_size(self, balance: float) -> float:
        """Get position size from Treasury Vault block."""
        for block in self.blocks:
            if block.get("id") == "position-size":
                pct = block.get("params", {}).get("percentage", 10)
                return balance * (pct / 100)
        return balance * 0.1

    def check_exit_conditions(
        self, entry_price: float, current_price: float
    ) -> Tuple[bool, Optional[str], float]:
        """Check Stop-Loss Tower, Profit Beacon, and Guardian Moat."""
        pnl_pct = ((current_price - entry_price) / entry_price) * 100

        for block in self.blocks:
            block_id = block.get("id", "")
            params = block.get("params", {})

            if block_id == "stop-loss":
                if pnl_pct <= -params.get("percentage", 2):
                    return True, "stop_loss", pnl_pct

            elif block_id == "take-profit":
                if pnl_pct >= params.get("percentage", 5):
                    return True, "take_profit", pnl_pct

            elif block_id == "trailing-stop":
                if pnl_pct <= -params.get("percentage", 3):
                    return True, "trailing_stop", pnl_pct

        return False, None, pnl_pct

    def execute_step(self, market_data: pd.DataFrame, balance: float) -> Dict:
        """Execute one trading step."""
        if len(market_data) < 2:
            return {"action": "HOLD", "message": "Gathering market intel..."}

        current_price = float(market_data["close"].iloc[-1])

        # Check entry
        if not self.positions and self.should_enter(market_data):
            position_size = self.calculate_position_size(balance)
            self.positions.append(
                {
                    "entry_price": current_price,
                    "size": position_size,
                    "timestamp": str(market_data.index[-1]),
                }
            )
            self.trade_count += 1

            return {
                "action": "ENTER",
                "price": current_price,
                "size": position_size,
                "message": "Gates opened! Entering position...",
            }

        # Check exit
        if self.positions:
            pos = self.positions[0]
            should_exit, reason, pnl_pct = self.check_exit_conditions(
                pos["entry_price"], current_price
            )

            if should_exit:
                profit = pos["size"] * (pnl_pct / 100)
                self.pnl += profit
                self.positions = []

                msg = f"{reason.replace('_', ' ').title()}! P&L: ${profit:.2f}"

                return {
                    "action": "EXIT",
                    "reason": reason,
                    "price": current_price,
                    "pnl": profit,
                    "message": msg,
                }

        return {"action": "HOLD", "message": "Holding strong..."}

    def _calculate_rsi(self, data: pd.DataFrame, period: int = 14) -> float:
        """Calculate RSI indicator."""
        if len(data) < period + 1:
            return 50.0

        delta = data["close"].diff()
        gain = delta.where(delta > 0, 0).rolling(window=period).mean()
        loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()

        rs = gain / loss
        rsi = 100 - (100 / (1 + rs))
        val = rsi.iloc[-1]

        return float(val) if not np.isnan(val) else 50.0

File: src/game/strategyEngine.js
/**
 * Client-side strategy evaluation engine.
 * Converts block configurations into trading logic.
 */

export function calculateRSI(closes, period = 14) {
  if (closes.length < period + 1) return 50;

  let gains = 0;
  let losses = 0;

  for (let i = closes.length - period; i < closes.length; i++) {
    const delta = closes[i] - closes[i - 1];
    if (delta > 0) gains += delta;
    else losses -= delta;
  }

  const avgGain = gains / period;
  const avgLoss = losses / period;

  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function calculateSMA(closes, period) {
  if (closes.length < period) return null;
  const slice = closes.slice(-period);
  return slice.reduce((a, b) => a + b, 0) / period;
}

export function calculateMACD(closes, fast = 12, slow = 26, signal = 9) {
  if (closes.length < slow + signal) return { macd: 0, signal: 0, histogram: 0 };

  const emaFast = calculateEMA(closes, fast);
  const emaSlow = calculateEMA(closes, slow);
  const macdLine = emaFast - emaSlow;

  return { macd: macdLine, signal: 0, histogram: macdLine };
}

function calculateEMA(data, period) {
  if (data.length < period) return data[data.length - 1];

  const k = 2 / (period + 1);
  let ema = data.slice(0, period).reduce((a, b) => a + b, 0) / period;

  for (let i = period; i < data.length; i++) {
    ema = data[i] * k + ema * (1 - k);
  }

  return ema;
}

export function evaluateEntryConditions(blocks, candles) {
  const closes = candles.map((c) => c.close);
  if (closes.length < 2) return false;

  const entryBlocks = blocks.filter((b) => b.category === 'entry');
  if (entryBlocks.length === 0) return false;

  for (const block of entryBlocks) {
    switch (block.id) {
      case 'rsi-gate': {
        const rsi = calculateRSI(closes);
        if (rsi >= block.params.threshold) return false;
        break;
      }
      case 'ma-cross': {
        const fastMA = calculateSMA(closes, block.params.fast);
        const slowMA = calculateSMA(closes, block.params.slow);
        if (!fastMA || !slowMA || fastMA <= slowMA) return false;
        break;
      }
      case 'macd-signal': {
        const { macd } = calculateMACD(closes, block.params.fast, block.params.slow, block.params.signal);
        if (macd <= 0) return false;
        break;
      }
      default:
        break;
    }
  }

  return true;
}

export function getPositionSize(blocks, balance) {
  const sizingBlock = blocks.find((b) => b.id === 'position-size');
  const pct = sizingBlock ? sizingBlock.params.percentage : 10;
  return balance * (pct / 100);
}

export function checkExitConditions(blocks, entryPrice, currentPrice) {
  const pnlPct = ((currentPrice - entryPrice) / entryPrice) * 100;

  for (const block of blocks) {
    if (block.id === 'stop-loss' && pnlPct <= -block.params.percentage) {
      return { shouldExit: true, reason: 'stop_loss', pnlPct };
    }
    if (block.id === 'take-profit' && pnlPct >= block.params.percentage) {
      return { shouldExit: true, reason: 'take_profit', pnlPct };
    }
    if (block.id === 'trailing-stop') {
      // Simplified trailing stop
      if (pnlPct <= -block.params.percentage) {
        return { shouldExit: true, reason: 'trailing_stop', pnlPct };
      }
    }
  }

  return { shouldExit: false, reason: null, pnlPct };
}
